llvm/lib/CodeGen/CalcSpillWeights.cpp

```diff
@@ -146,6 +146,17 @@ void VirtRegAuxInfo::calculateSpillWeightAndHint(LiveInterval &LI) {
   LI.setWeight(Weight);
 }
 
+static bool canMemFoldInlineAsm(LiveInterval &LI,
+                                const MachineRegisterInfo &MRI) {
+  for (const MachineOperand &MO : MRI.reg_operands(LI.reg())) {
+    const MachineInstr *MI = MO.getParent();
+    if (MI->isInlineAsm() && MI->mayFoldInlineAsmMemOp(MI->getOperandNo(&MO)))
+      return true;
```

Comment:
What happens if a register is both foldable and not foldable?
I.e., something like `asm "rm, r", %0, %0`.

Given the second constraints force `%0` to stay in register, we can't fold right.

In other words, depending on the semantic that we want, we may want to check all the operands.
Now, I still approved this PR because although the resulting `LI` won't actually be spillable in this case, regalloc will still handle it correctly.

Reply:
> What happens if a register is both foldable and not foldable?
> I.e., something like asm "rm, r", %0, %0.

I'm not sure that can be expressed from inline asm.

Do you mean something like:
```c
asm ("# %0 %1"::"rm"(x), "r"(x));
```
?

I /that/ case, `%0` would have a distinct storage location from `%1`. `x` would be copied into the inline asm twice.  `%0`'s constraint is `rm` (register /OR/ memory), `%1`'s constraint is `r` (register).  I don't think AND NOT is expressible via the constraint language.